noria-server/mir/src/optimize.rs

```diff
@@ -1,19 +1,217 @@
+    // of the aggregate, in which case they cannot be merged.
+
+    let mut candidate_nodes = Vec::new();
+
+    for n in found_nodes {
+        // if we've already set this to false, we shouldn't look at it more;
+        // e.g. it might be merging with its parent node.
+        let node_name = n.borrow().versioned_name();
+        if !visited_nodes.get(&node_name).unwrap() {
+            continue;
+        }
+        // now scan for candidacy
+        let mut candidate = false;
+        match n.borrow().inner {
+            MirNodeType::Filter { .. }  => {
+                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // then this is a candidate
+                if n.borrow().children.len() == 1 {
+                    let temp = n.borrow();
+                    let child = temp.children.first().unwrap().borrow();
+                    match child.inner {
+                        MirNodeType::Aggregation { .. } => {
+                            if child.ancestors.len() == 1 {
+                                candidate = true;
+                            }
+                        },
+                        _ => {},
+                    };
+                }
+            },
+            MirNodeType::Aggregation { ref on, .. }  => {
+                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // then this is a candidate
+                if n.borrow().children.len() == 1 {
+                    let temp = n.borrow();
+                    let child = temp.children.first().unwrap().borrow();
+                    match child.inner {
+                        MirNodeType::Filter { ref conditions } => {
+                            if child.ancestors.len() == 1 {
+                                candidate = true;
+
+                                // But wait -- need to check if the filter is on the aggregation result
+                                use nom_sql::FunctionExpression::{Count, Sum};
+                                for (i, col) in child.columns.iter().enumerate() {
+                                    match col.function.clone() {
+                                        Some(func_expr) => {
+                                            match *func_expr {
+                                                 Count(col, _)
+                                                 | Sum(col, _) => {
+                                                    if col.name == on.name {
+                                                        // this column may be the aggregation result
+                                                        // so if we're filtering on it, we're not a candidate
+                                                        for (j, _cond) in conditions {
+                                                            if *j == i {
+                                                                candidate = false;
```

Comment:
Do we want to break early if this is the case to avoid looking at a bunch of other conditions/children?

Reply:
Sounds good.